File: sizzle/lib/entities/product.dart
class Product {
  final String id;
  final String name;
  final String description;
  final List<String> imageUrls;
  final int categoryId;
  final String category;
  final int ratingsCount;
  final String brand;
  final int price;
  final double ratings;
  final int quantity;
  final DateTime dateAdded;
  final double discountPercentage;
  final List<String> reviews;
  final String videoUrl;

  Product({
    required this.id,
    required this.name,
    required this.description,
    required this.imageUrls,
    required this.categoryId,
    required this.category,
    required this.ratingsCount,
    required this.brand,
    required this.price,
    required this.ratings,
    required this.quantity,
    required this.dateAdded,
    required this.discountPercentage,
    required this.reviews,
    required this.videoUrl,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      imageUrls: List<String>.from(json['imageUrls']),
      categoryId: json['categoryId'],
      category: json['category'],
      ratingsCount: json['ratings_count'],
      brand: json['brand'],
      price: json['price'],
      ratings:
          json['ratings']?.toDouble() ?? 0.0, // Ensures null safety for ratings
      quantity: json['quantity'],
      dateAdded: DateTime.parse(json['date_added']),
      discountPercentage: json['discount_percentage']?.toDouble() ??
          0.0, // Ensures null safety for discount percentage
      reviews: List<String>.from(json['reviews']),
      videoUrl: json['video_url'],
    );
  }
}

File: backend/products/src/main/java/com/barniq/products/entities/Product.java
package com.barniq.products.entities;

import com.barniq.products.config.AttributeConverterClass;
import com.vladmihalcea.hibernate.type.json.JsonBinaryType;
import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;


import com.vladmihalcea.hibernate.type.json.JsonBinaryType;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;


@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "products")

public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;
    private String description;

    @ManyToOne
    @JoinColumn(name = "category_id")
    private Category category;

    private int ratingsCount;
    private String brand;
    private int price;
    private double ratings;
    private int quantity;
    private LocalDateTime dateAdded;
    private double discountPercentage;
    private String videoUrl;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> attributes;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<ProductImage> imageUrls= new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<ProductReview> reviews= new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<ProductVariant> variants= new ArrayList<>();

    @Column(unique = true, nullable = false)
    private String slug;

    @PrePersist
    public void prePersist() {
        this.dateAdded = LocalDateTime.now();
    }



}
